src/python/gudhi/representations/metrics.py

```diff
@@ -10,17 +10,114 @@
 import numpy as np
 from sklearn.base import BaseEstimator, TransformerMixin
 from sklearn.metrics import pairwise_distances
-try:
-    from .. import bottleneck_distance
-    USE_GUDHI = True
-except ImportError:
-    USE_GUDHI = False
-    print("Gudhi built without CGAL: BottleneckDistance will return a null matrix")
+from gudhi.hera import wasserstein_distance as hera_wasserstein_distance
+from .preprocessing import Padding
 
 #############################################
 # Metrics ###################################
 #############################################
 
+def sliced_wasserstein_distance(D1, D2, num_directions):
+    """
+    This is a function for computing the sliced Wasserstein distance from two persistence diagrams. The Sliced Wasserstein distance is computed by projecting the persistence diagrams onto lines, comparing the projections with the 1-norm, and finally integrating over all possible lines. See http://proceedings.mlr.press/v70/carriere17a.html for more details.
+    :param D1: (n x 2) numpy.array encoding the (finite points of the) first diagram. Must not contain essential points (i.e. with infinite coordinate).
+    :param D2: (m x 2) numpy.array encoding the second diagram.
+    :param num_directions: number of lines evenly sampled from [-pi/2,pi/2] in order to approximate and speed up the distance computation.
+    :returns: the sliced Wasserstein distance between persistence diagrams.
+    :rtype: float 
+    """
+    thetas = np.linspace(-np.pi/2, np.pi/2, num=num_directions+1)[np.newaxis,:-1]
+    lines = np.concatenate([np.cos(thetas), np.sin(thetas)], axis=0)
+    approx1 = np.matmul(D1, lines)
+    diag_proj1 = (1./2) * np.ones((2,2))
+    approx_diag1 = np.matmul(np.matmul(D1, diag_proj1), lines)
+    approx2 = np.matmul(D2, lines)
+    diag_proj2 = (1./2) * np.ones((2,2))
+    approx_diag2 = np.matmul(np.matmul(D2, diag_proj2), lines)
+    A = np.sort(np.concatenate([approx1, approx_diag2], axis=0), axis=0)
+    B = np.sort(np.concatenate([approx2, approx_diag1], axis=0), axis=0)
+    L1 = np.sum(np.abs(A-B), axis=0)
+    return np.mean(L1)
```

Comment:
If I understand this code correctly: it used to be that we could precompute some things about diagrams in `fit` (their projections along each line, and the projections along each line of their diagonal projections, possibly pre-sorted to speed up further sorting). But now we have to recompute that for each diagram as many times as there are other diagrams. I don't know how important it is for this specific distance, but in general it doesn't look so nice. I think those computations should be done in 2 steps, one where we preprocess diagrams (say replace a diagram by all the projections), and then the step where we compute some quantity on each pair of **preprocessed** diagrams. Does that make sense?

Reply:
Yes that totally makes sense. I kept the `sliced_wasserstein_distance` function, but I also added `sliced_wasserstein_distance_on_projections` and `compute_persistence_diagram_projections` to speed up the computation of all pairwise distances.